provisioning/main.c

```diff
@@ -0,0 +1,345 @@
+        goto done;
+    }
+
+    mbedtls_x509write_csr_set_key_usage(
+        &req, MBEDTLS_X509_KU_DIGITAL_SIGNATURE);
+
+    mbedtls_x509write_csr_set_ns_cert_type(
+        &req, MBEDTLS_X509_NS_CERT_TYPE_SSL_CLIENT);
+
+    mbedtls_x509write_csr_set_key(&req, &pk);
+
+    ret = mbedtls_x509write_csr_pem(&req, output, CSR_PEM_LEN,
+        psa_rng_for_mbedtls, NULL);
+    if (ret != 0)
+    {
+        printf("Failed to make CSR with ret=%d\n", ret);
+        goto done;
+    }
+
+    printf("\tKey ID %lu:\n", key_id);
+    printf("%s\n", output);
+
+done:
+    free(output);
+    mbedtls_x509write_csr_free(&req);
+    mbedtls_pk_free(&pk);
+}
+
+int main(void)
+{
+    psa_status_t status;
+    printf("Provisioning device...\n");
+
+    /* Erase secure storage (PSA ITS and PSA PS) to give Mbed Crypto a blank
+     * slate to provision the device from. Any keys it used to know about will
+     * be forgotten. Any keys stored in PSA ITS will be erased. All metadata
+     * stored in PSA ITS about keys stored in secure elements will be erased.
+     *
+     * Note that the device may reboot as a part of this erase step, depending
+     * on how mbed_psa_reboot_and_request_new_security_state() is implemented.
+     * Currently, the function does not reboot in a manner such that this
+     * function would be executed again and again in a loop.
+     */
+    printf("\tErasing secure storage... ");
+    fflush(stdout);
+    status = mbed_psa_reboot_and_request_new_security_state(PSA_LIFECYCLE_ASSEMBLY_AND_TEST);
+    if (status != PSA_SUCCESS)
+    {
+        printf("failed with status=%ld\n", status);
+        return status;
+    }
+    printf("done.\n");
+
+    printf("\tRegistering drivers... ");
+    fflush(stdout);
+    status = psa_register_se_driver(PSA_ATECC608A_LIFETIME, &atecc608a_drv_info);
+    if (status != PSA_SUCCESS)
+    {
+        printf("failed with status=%ld\n", status);
+        return status;
+    }
+    printf("done.\n");
+
+    printf("\tInitializing PSA Crypto... ");
+    fflush(stdout);
+    status = psa_crypto_init();
+    if (status != PSA_SUCCESS)
+    {
+        printf("failed with status=%ld\n", status);
+        return status;
+    }
+    printf("done.\n");
+
+    /* The secure element factory put a device private key pair into a slot in
+     * the secure element. We need to tell Mbed Crypto that this key pair
+     * exists so that it can be used. */
+    printf("\tRegistering factory-created keys... ");
+    fflush(stdout);
+    status = register_preprovisioned_keys();
+    if (status != PSA_SUCCESS)
+    {
+        printf("failed with status=%ld\n", status);
+        return status;
+    }
+    printf("done.\n");
+
+    /* We also want to generate our own key entirely within the secure element
+     * during the factory device provisioning stage in device assembly. Ask
+     * Mbed Crypto to generate this key for us. */
+    printf("\tGenerating keys within the secure element... ");
```

Comment:
Not a blocker, but it's a bit weird that the message says “generating key**s**” but the function is `generate_key` singular. And the registration above uses plural both in the message and in the function name, but actually only registers one key (ok, technically a key pair, but technically a key pair is a single key object).